### PR TITLE
Get PIDs from cgroups recursively

### PR DESCRIPTION
**Pull request by LK4D4**
Closes #331
Closes #329
@ibuildthecloud I carried and slightly changed code. Now you can just review and not spend time on my comments in your branch.

### CODE REVIEW

File: libcontainer/cgroups/fs/apply_raw.go [hunk 1]
```diff
@@ -213,7 +213,24 @@ func (m *Manager) GetPids() ([]int, error) {
 		return nil, err
 	}
 
-	return cgroups.ReadProcsFile(dir)
+	var pids []int
+	var readErr error
+	// collect pids from all sub-cgroups
+	filepath.Walk(dir, func(p string, info os.FileInfo, err error) error {
```
Comment (crosbymichael) [inline libcontainer/cgroups/fs/apply_raw.go]: u are not getting the error from `Walk` here?

Comment (LK4D4) [inline libcontainer/cgroups/fs/apply_raw.go]: Nope, I'm not returning it as well. I think I need to return at leas in case of "cgroup.procs"

Comment (LK4D4) [inline libcontainer/cgroups/fs/apply_raw.go]: Idea was to get as much pids as possible :)

**Comment by LK4D4** [comment]
@crosbymichael I changed to return any error on procs file processing.

File: libcontainer/cgroups/fs/apply_raw.go [hunk 2]
```diff
@@ -213,7 +213,27 @@ func (m *Manager) GetPids() ([]int, error) {
...
+	var pids []int
+	// collect pids from all sub-cgroups
+	err = filepath.Walk(dir, func(p string, info os.FileInfo, iErr error) error {
+		if info.IsDir() {
+			return nil
+		}
+
+		dir, file := filepath.Split(p)
+		if file == "cgroup.procs" {
+			if iErr != nil {
```
Comment (crosbymichael) [inline libcontainer/cgroups/fs/apply_raw.go]: why is this check here?  shouldn't it be at the top?

Comment (LK4D4) [inline libcontainer/cgroups/fs/apply_raw.go]: I don't want to stop process on errors for other files. In user-namespace this can be an issue.

Comment (crosbymichael) [inline libcontainer/cgroups/fs/apply_raw.go]: humm, interesting...

But there error that you get in `iErr` happens from the previous iteration right?

Comment (tiborvass) [inline libcontainer/cgroups/fs/apply_raw.go]: @crosbymichael if you're talking about the walk iteration, `iErr` is the error for the current path or a file directly under the current path.

Comment (LK4D4) [inline libcontainer/cgroups/fs/apply_raw.go]: yes, I think `iError` is error from `Open` on `p`.

**Comment by hqhq** [comment]
Don't we need to change `apply_systemd.go` as well?

**Comment by ibuildthecloud** [comment]
@LK4D4 thanks, we will test out this patch.

**Comment by LK4D4** [comment]
@hqhq You're right, I'll fix tomorrow. It's weird that implementations are different.

**Comment by LK4D4** [comment]
@hqhq Updated for systemd.

File: libcontainer/cgroups/systemd/apply_systemd.go [hunk 3]
```diff
@@ -416,7 +416,27 @@ func (m *Manager) GetPids() ([]int, error) {
 		return nil, err
 	}
 
-	return cgroups.ReadProcsFile(path)
+	var pids []int
```
Comment (mrunalp) [inline libcontainer/cgroups/systemd/apply_systemd.go]: Can we move the code into a common functions in utils to avoid duplication?

Comment (LK4D4) [inline libcontainer/cgroups/systemd/apply_systemd.go]: Let's do this, what do you think about same cgroup for both implementations? Like "devices".

Comment (mrunalp) [inline libcontainer/cgroups/systemd/apply_systemd.go]: sounds good.

**Comment by crosbymichael** [comment]
LGTM

**Comment by dqminh** [comment]
LGTM

**Comment by tiborvass** [comment]
IANTM but `make test` passed for me after merging it in master locally.

**Comment by mrunalp** [comment]
What's up with janky?

**Comment by mrunalp** [comment]
Closes #331

**Comment by tiborvass** [comment]
@mrunalp the "Closes #XXX" has to be part of the description of the PR. (you can edit it)

Janky is a little sick, you need to test manually.

**Comment by ibuildthecloud** [comment]
We are having some issues with this patch, still investigating.

**Comment by ibuildthecloud** [comment]
@lk4d4 ^^

**Comment by tiborvass** [comment]
@ibuildthecloud there's a panic being fixed, see docker/docker#16989

**Comment by ibuildthecloud** [comment]
Anybody want to try my original patch?

**Comment by tiborvass** [comment]
@ibuildthecloud feel free to review https://github.com/opencontainers/runc/pull/332

**Pull request merged**